internal/mode/static/state/store.go

```diff
@@ -219,23 +220,21 @@ func (s *changeTrackingUpdater) Upsert(obj client.Object) {
 func (s *changeTrackingUpdater) delete(objType client.Object, nsname types.NamespacedName) (changed bool) {
 	objTypeGVK := s.extractGVK(objType)
 
-	if !s.persistedGVKs.contains(objTypeGVK) {
-		return false
-	}
+	// persistedGVKs will only contain objTypeGVK if the resource is being tracked in the store.
+	if s.persistedGVKs.contains(objTypeGVK) {
+		if s.store.get(objType, nsname) == nil {
+			return false
+		}
 
-	obj := s.store.get(objType, nsname)
-	if obj == nil {
-		return false
+		s.store.delete(objType, nsname)
 	}
 
-	s.store.delete(objType, nsname)
-
 	stateChanged, ok := s.stateChangedPredicates[objTypeGVK]
 	if !ok {
 		return false
 	}
 
-	return stateChanged.delete(obj)
+	return stateChanged.delete(objType)
```

Comment:
It's not the same. 

`objType` is the concrete type of the resource. For example, `*v1.HTTPRoute`. However, it's empty, the equivalent of `&v1.HTTPRoute{}`. 

The `obj` from the store is also a `*v1.HTTPRoute`, but it is non-empty. 

In this case, we want to use `obj` because we need to extract the nsname of the object in order to check if it is referenced in the latest graph. 

Alternatively, the `delete` method on `stateChangedPredicate` interface could be changed to: 
```
delete(objType client.Object, nsname types.NamespacedName) bool
```

Then, you could pass the `objType` instead of `obj`. This is probably a clearer method signature.

Reply:
Ahhhh 😦 

Thinking of the scenario of an `EndpointSlice` resource which is not stored in the store -- I'm having difficulty thinking of a way to get the `nsname` of the object passed down as `obj := s.store.get(objType, nsname)` will return nil (or panic) since its not in the store.

My initial thought of fixing this is to refactor such that `delete` takes an entire Resource instead of just the objType. Is there any other way?

Reply:
The `nsname ` is passed into the delete function on `changeTrackingUpdater`, so you already have it. 

If you change the `delete` method on the `stateChangedPredicate` interface and update all the implementations to accept `objType` and `nsname`, then this should work:

```
objTypeGVK := s.extractGVK(objType)

if s.persistedGVKs.contains(objTypeGVK) {
    s.store.delete(objType, nsname)
}

stateChanged, ok := s.stateChangedPredicates[objTypeGVK]
if !ok {
    return false
}

return stateChanged.delete(objType, nsname)
```

In this solution you don't need to get the obj from the store because you already have the objType and nsname. You just need to check if the GVK is persisted so you can delete the object in the store.

Reply:
🤦 👍